### ISSUE TITLE
[Question] Is there any way to protect stream publish URL?

### ISSUE THREAD

**Issue by EliotBerriot** [ant-media/Ant-Media-Server#579]
Hi there, I don't know if it's the proper place to ask this kind of questions, let me know if you'd rather have this published somewhere else.

I'm currently doing some research on media streaming servers to integrate live streaming into [Funkwhale](https://funkwhale.audio), the project I'm working on.

Ant media server looks like a really good pick for my use cases, but I'm not sure how to ensure viewers cannot hijack a stream.

When creating a Stream in the interface, you can:

- Copy the RTMP publish url, like rtmp://localhost/LiveApp/499481361945988697107161, `499481361945988697107161` being the Stream ID
- Copy the player embbed code, like `<iframe src="//localhost:5080/LiveApp/play.html?name=499481361945988697107161" frameborder="0" allowfullscreen></iframe>`, `499481361945988697107161` being the stream ID, the same one as in the publish URL

Since the Stream ID is used for publishing purposes, but also shared with viewers, my understanding is that any viewer can easily guess the publish URL by inspecting the player source and potentially hijack a stream. I may be completely wrong or missing something, but I'd like to double check that with you.

Is this normal behaviour? Is there another way to achieve what I need (having a non guessable rmtp publishing link)? I can see that there is another issue linked to authentication (https://github.com/ant-media/Ant-Media-Server/issues/138), but my idea was that providing a different, non guessable URL (with a different ID, but binded to the same stream) for viewing / publishing.

Let me know if you need additional info, and thank you for the incredible work!

**Comment by davutcavdar** [comment]
@EliotBerriot 
Yes, your mentioned use case is true, publishing and playing are performed according to the unique stream Id. For solve security concerns, Token Control mechanism is developed for the Enterprise Edition. This solution protects your publisihing and playing operations. Please have a look at these wiki and blog pages. If you need further assistance please send an email to contact[at]antmedia.io.

Blog : https://antmedia.io/secure-video-streaming/
Wiki :https://github.com/ant-media/Ant-Media-Server/wiki/The-Steps-of-One-Time-Token-Operations

**Comment by EliotBerriot** [comment]
@davutcavdar thank you for your answer, which raises some concerns.

I understand that people need to earn a living, hence the community/enterprise edition split. However, in most if not all projects, the community edition implement the core features that are absolutely mandatory, while the enterprise edition add *extra* features that are needed for more advanced use cases or bigger installations.

The way I understand it, a basic media streaming server allows:

1. One user to publish a stream
2. Everyone to receive a stream

Item 1 is simply not available in the community edition, since everyone can actually publish on *any* stream. In some cases, *additional* security is an *extra* feature, but that's not true here: having an even basic restriction on who can publish on a given stream is part of the core use case. Without that, the software is actually dangerous to use for any real-world stream!

Let me take a concrete example showing how bad it could be.

Alice has a video game a stream on an Ant Media Server (Community edition), with some young viewers in the audience. Alice's streams are completely safe for any audience. But one day, for fun, someone decides to hijack Alice's stream and broadcast porn instead. *They can do that in literally 1 minute, by simply examining the Stream URL in the page source and firing up OBS*. Alice looses her audience, reputation, and can even be sued by angry parents.

Knowing that publishing is left completely open, I don't understand why anyone would use the Community edition of Ant Media. The attack surface and risks are just too big.

I sincerely think publishing should be protected by default in the community edition, and I hope that you'll consider that for the safety of viewers and streamers on your platform.

**Comment by davutcavdar** [comment]
Thanks for your feedback. They are valuable for us. 

We will think about it with the team, make some evaluation and let you know.

**Comment by EliotBerriot** [comment]
Hi @davutcavdar, any news from the team?

**Comment by davutcavdar** [comment]
Hi @EliotBerriot,

Sorry for late response :( We discussed with the team and decided to continue token control in Enterprise Edition. But we can support you about dealing with this issue with Community Edition.

You can rename produced files (HLS m3u8 or Mp4) using Ant Media Server Muxers 
Init methods (https://github.com/ant-media/Ant-Media-Server-Common/tree/master/src/main/java/io/antmedia/muxer)

Just an example, for HLS Muxer go to Init method and edit this line to change output file names: https://github.com/ant-media/Ant-Media-Server-Common/blob/836a4fbb1955f43b456d69186490d1378350fd4c/src/main/java/io/antmedia/muxer/HLSMuxer.java#L150 

By doing that, users can not guess stream publish id by analyzing stream URL. Please let us know if you have any issue or just send an email.

**Comment by EliotBerriot** [comment]
Hi @davutcavdar, thank you for reaching back to me

I have to say I am a bit surprised by your team answer though.

Did you consider the security risks associated with the current state of your software? If all Community Edition deployments and Enterprise Edition streams that do not use the Token Control feature are vulnerable to stream hijacking, I believe you have a responsability to fix that. 

Your current stance is "our software is insecure, but you can enable a paid option to make it secure". And even when paying, the token thing is still optional! That's simply not acceptable, especially since you're not advertising this huge security risk anywhere. 

We're not talking about enhanced, nice-to-have security here (like with two-factor authentication which could be a paid option only), it's about basic, indispensable security for publishers and viewers!

Thank you for the code samples, but there is no point in implementing a workaround to secure your software if you're not planning to secure it yourself in the end. If it's really that simple, to fix, I think your team should take an hour or two to implement this fix and protect _all_ the people using your software.

**Comment by jomo** [comment]
So basically you have an extremely easy to exploit security vulnerability and decided to only fix this is the paid Enterprise Edition?

**Comment by davutcavdar** [comment]
@EliotBerriot thanks for feedbacks. 
I should correct some points because it causes some misunderstandings. First of all Community edition provides more and more features even includes some enterprise features and used by lots of developers in the community.

Using unique stream id both for streaming and playing is just about implementation example and you can manage other things on your application level based on this open-source project. I also gave an example about how you can change output name of streaming files that you will give to users. Therefore changing these names, users can not reach your stream (Alice's stream ) URL and you do not have any security threat at this point.

To make it clear, One-time token implementation is developed for only limiting audience not for fixing a security issue.

To sum up, Ant Media Server Community Edition is totally open source software and you can develop your application based on that, during these phases, you can change the structure, way of operation etc. Sure for them, you can get support, documentation etc every time.
